<?xml version="1.0"?>
<!--
Copyright (c) Aalto  - All Rights Reserved
Created on: 11/13/18
    Author: Vladimir Petrik <vladimir.petrik@aalto.fi>
   Details:
-->

<launch>

    <arg name="look_at_x" default="0.0"/>
    <arg name="look_at_y" default="0.0"/>
    <arg name="look_at_z" default="-0.051"/>

    <arg name="rviz" default="true"/>

    <arg name="model" default="$(find lumi_description)/robots/robot.urdf"/>
    <arg name="mujoco_model" default="$(arg model)"/>

    <!--Compensate bias forces in mujoco; If true the effort visible to ROS will not contains gravity/corollios/etc. forces -->
    <arg name="compensate_bias" default="false"/>
    <arg name="bias_error" default="1.0"/>

    <arg name="robot_name" default="lumi"/>

    <param name="robot_description" textfile="$(arg model)"/>

    <group ns="lumi_mujoco">
        <rosparam command="load" file="$(find lumi_mujoco)/config/controllers.yaml"/>
        <arg name="controllers" default="joint_state_controller trajectory_controller trajectory_controller_fingers"/>
        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"
              args="$(arg controllers)"/>

        <node pkg="topic_tools" type="relay" name="relay_joint_state" args="joint_states /joint_states"/>
    </group>

    <node name="lumi_mujoco" type="mujoco_control" pkg="mujoco_ros_control" output="screen">
        <param name="model" value="$(arg mujoco_model)"/>
        <remap from="/lumi_mujoco/trajectory_controller/follow_joint_trajectory"
               to="/lumi_arm_controller/follow_joint_trajectory"/>
        <remap from="/lumi_mujoco/trajectory_controller_fingers/follow_joint_trajectory"
               to="/lumi_hand_controller/follow_joint_trajectory"/>
        <param name="lumi_joint4_init" value="-1.5708"/>
        <param name="lumi_joint6_init" value="1.5708"/>
        <param name="lumi_finger_joint1_init" value="0.02"/>
        <param name="lumi_finger_joint2_init" value="0.02"/>
        <param name="look_at_x" value="$(arg look_at_x)"/>
        <param name="look_at_y" value="$(arg look_at_y)"/>
        <param name="look_at_z" value="$(arg look_at_z)"/>
        <param name="compensate_bias" value="$(arg compensate_bias)"/>
        <param name="bias_error" value="$(arg bias_error)"/>
    </node>

    <node name="lumi_gripper_action_forwarding" type="gripper_action_to_trajectory_action" pkg="lumi_mujoco"
          output="screen">
	<param name="robot_name" value="$(arg robot_name)"/>
        <remap from="/lumi_gripper_action_forwarding/gripper_action" to="/lumi_hand_controller/gripper_action"/>
        <remap from="/lumi_gripper_action_forwarding/follow_joint_trajectory"
               to="/lumi_hand_controller/follow_joint_trajectory"/>
    </node>

    <!--Load Move Group Interface -->
    <include file="$(find lumi_moveit_config)/launch/lumi_moveit.launch">
        <arg name="rviz" value="$(arg rviz)"/>
    </include>

</launch>
